enhancements/sig-architecture/229-manifestworkreplicaset-rollback/README.md

```diff
@@ -0,0 +1,436 @@
+
+On the other hand, `Undo` is the explicit action. If user wants to undo to the specific old revision, the user (or CLI) can find the revision from the list of ControllerRevision resources and apply the older manifestTemplate to update the `.spec.manifestWorkTemplate`. Updated MWRS executes the normal rollout process based on rolloutStrategy configuration.
+
+#### Enhanced reconcile logic
+
+1. `(NEW)` Reconstruct ManifestWorkReplicaSet history:
+    1. List existing `ControllerRevision`s controlled by this `ManifestWorkReplicaSet`.
+    1. Find the history of the current `ManifestWorkReplicaSet`'s `.spec.ManifestWorkTemplate` desired state, and create one if not found:
+    1. The `.name` of the history will be unique, generated from `.spec.ManifestWorkTemplate` hash with hash collision resolution (`.status.collisionCount`) 
+    1. If history creation failed:
+        1. If it is because of name collision:
+            1. If the collided history is same as `ManifestWorkReplicaSet`'s `.spec.ManifestWorkTemplate` desired state, there is the already created the history
+            1. Otherwise, bump `ManifestWorkReplicaSet`'s `.status.collisionCount` by 1, 
+        1. Else, exit and reconcile again
+    1. New ControllerRevision will be labeled with `work.open-cluster-management.io/controller-hash`
+    1. Work controller will add a ControllerRef in the history `.ownerReferences`.
+    1. The current history should be the largest `.revision` number amongst the others. Update `.revision` to the largest if it is not. (for example, after rollback, we will need to update it to the latest.)
+2. `(NEW)` Update `.spec.updateRevision` to the newly calculated revision.
+3. Iterate each placement references (`.spec.placementRefs`)
+    1. Get all manifestworks associated with the current placement
+    1. `(NEW)` Find the progressing manifestwork hash
+        1. Use `.spec.updateRevision` if `.status.abort` is true, otherwise, use `.status.currentRevision`.
+    1. Find manifestworks which has the same hash of the progressing manifestwork hash.
+        1. [Determine cluster rollout status (Failed, Success)](https://github.com/open-cluster-management-io/ocm/blob/main/pkg/work/hub/controllers/manifestworkreplicasetcontroller/manifestworkreplicaset_deploy_reconcile.go#L96)
+    1. Create Rollout handler
+    1. Calculate RolloutResult (rollout/timeout/removed candidate clusters)
+    1. `(NEW)` If rolloutResult includes timeout clusters and `.spec.placementRefs[*].rolloutStrategy.abortOnFailure` is true
```

Comment:
What's the condition to trigger automatic abort? How about the failed clusters?

Reply:
When minSuccessTime is exceeded for the current rollout group, it will start the abort process.

Reply:
@haoqing0110 I added new step `Evaluate abort condition` Please review it to see if it makes sense.